pkg/reconciler/configrevision_reconciler.go

```diff
@@ -0,0 +1,466 @@
+	revisions, err := listRevisions(ctx, crr.client)
+	if err != nil {
+		return fmt.Errorf("error listing revisions: %w", err)
+	}
+
+	if len(revisions.Items) > 1 {
+		nodeConfigs, err := crr.listConfigs(ctx)
+		if err != nil {
+			return fmt.Errorf("error listing configs: %w", err)
+		}
+		if !revisions.Items[0].Status.IsInvalid && revisions.Items[0].Status.Ready == revisions.Items[0].Status.Total {
+			for i := 1; i < len(revisions.Items); i++ {
+				if countReferences(&revisions.Items[i], nodeConfigs.Items) == 0 {
+					if err := crr.client.Delete(ctx, &revisions.Items[i]); err != nil {
+						return fmt.Errorf("error deletring revision %s: %w", revisions.Items[i].Name, err)
+					}
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+func countReferences(revision *v1alpha1.NetworkConfigRevision, configs []v1alpha1.NodeNetworkConfig) int {
+	refCnt := 0
+	for j := range configs {
+		if configs[j].Spec.Revision == revision.Spec.Revision {
+			refCnt++
+		}
+	}
+	return refCnt
+}
+
+func (crr *ConfigRevisionReconciler) listConfigs(ctx context.Context) (*v1alpha1.NodeNetworkConfigList, error) {
+	nodeConfigs := &v1alpha1.NodeNetworkConfigList{}
+	if err := crr.client.List(ctx, nodeConfigs); err != nil {
+		return nil, fmt.Errorf("error listing nodeConfigs: %w", err)
+	}
+	return nodeConfigs, nil
+}
+
+func (crr *ConfigRevisionReconciler) deployNodeConfig(ctx context.Context, node *corev1.Node, revision *v1alpha1.NetworkConfigRevision) error {
+	currentConfig := &v1alpha1.NodeNetworkConfig{}
+	if err := crr.client.Get(ctx, types.NamespacedName{Name: node.Name}, currentConfig); err != nil {
+		if !apierrors.IsNotFound(err) {
+			return fmt.Errorf("error getting NodeNetworkConfig object for node %s: %w", node.Name, err)
+		}
+		currentConfig = nil
+	}
+
+	if currentConfig != nil && currentConfig.Spec.Revision == revision.Spec.Revision {
+		// current config is the same as current revision - skip
+		return nil
+	}
+
+	newConfig, err := crr.createConfigForNode(node, revision)
+	if err != nil {
+		return fmt.Errorf("error preparing config for node %s: %w", node.Name, err)
+	}
+
+	if err := crr.deployConfig(ctx, newConfig, currentConfig, node); err != nil {
+		if errors.Is(err, InvalidConfigError) || errors.Is(err, context.DeadlineExceeded) {
+			// revision results in invalid config or in context timeout - invalidate revision
+			revision.Status.IsInvalid = true
+			if err := crr.client.Status().Update(ctx, revision); err != nil {
+				return fmt.Errorf("error invalidating revision %s: %w", revision.Name, err)
+			}
+		}
+		return fmt.Errorf("error deploying config for node %s: %w", node.Name, err)
+	}
+	return nil
+}
+
+func (crr *ConfigRevisionReconciler) createConfigForNode(node *corev1.Node, revision *v1alpha1.NetworkConfigRevision) (*v1alpha1.NodeNetworkConfig, error) {
+	// create new config
+	c := &v1alpha1.NodeNetworkConfig{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: node.Name,
+		},
+	}
+
+	c.Spec = *revision.Spec.Config.DeepCopy()
```

Comment:
I don't think we need to deep copy here either, do we?

Reply:
You're right, changed that.